pkg/controller/registrysyncer/registrysyncer.go

```diff
@@ -216,17 +269,20 @@ func (r *reconciler) reconcile(ctx context.Context, req reconcile.Request, log *
 
 		*log = *log.WithField("imageStreamImport.Namespace", imageStreamImport.Namespace).WithField("imageStreamImport.Name", imageStreamImport.Name)
 		log.Debug("creating imageStreamImport")
-		if err := createAndCheckStatus(ctx, client, imageStreamImport); err != nil {
+		if err := r.createAndCheckStatus(ctx, key, client, imageStreamImport); err != nil {
 			controllerutil.CountImportResult(ControllerName, clusterName, req.Namespace, false)
-			return fmt.Errorf("failed to create and check the status for imageStreamImport on cluster %s %v : %w", clusterName, imageStreamImport, err)
+			return fmt.Errorf("failed to create and check the status for imageStreamImport for tag %s of %s in namespace %s on cluster %s: %w", imageTag, imageStreamImport.Name, imageStreamImport.Namespace, clusterName, err)
 		}
 		controllerutil.CountImportResult(ControllerName, clusterName, req.Namespace, true)
 		log.Debug("Imported successfully")
 	}
 	return nil
 }
 
-func createAndCheckStatus(ctx context.Context, client ctrlruntimeclient.Client, imageStreamImport *imagev1.ImageStreamImport) error {
+func (r *reconciler) createAndCheckStatus(ctx context.Context, key simpleImageStream, client ctrlruntimeclient.Client, imageStreamImport *imagev1.ImageStreamImport) error {
+	lock := r.imageStreamLocks.getLock(key)
```

Comment:
Please remove the `cluster` part from the lock and acquire/release it and the beginning/end of the reconciliation:
* It is much easier to understand
* It makes sure that we will either get to run the whole reconciliation or nothing, instead of possibly getting stuck waiting for the lock halfway through


A no-op reconciliation is extremely fast as its entirely based on in-memory data, so I don't think this will be a performance issue

Reply:
Tried this way while coding. Did not think this through. Cool.

Reply:
Done.